### ISSUE TITLE
Merging attributes

### ISSUE THREAD

**Issue by kuon** [mdgriffith/stylish-elephants#37]
When making an helper to apply attributes, I have to use it like so:

```
myHelper =
 [attr1, attr2]

el ([someAttr] ++ myHelper) empty
```

It would be very useful to have something like this:

```
myHelper =
 merge [attr1, attr2]

el [someAttr, myHelper] empty
```

For example, I am implementing tooltips, and I am thinking about something like this:

```
tooltip current name msg= 
 [ above (isCurrent current name) someTooltipElement
 , onMouseEnter (ShowTooltip name)
 , onMouseLeave (HideTooltip name)
 ]
```

```
el [attrs..., tooltip model.tooltip PlayButton "Play video"] ...
```

I don't know if the current attribute implementation permit this.

**Comment by spektroskop** [comment]
I've been experimenting with something like this here: https://github.com/spektroskop/stylish-elephants/commit/cc1fd22ff1078bb3dfaee69bf9307710439b27f8

I used the name ```mixin``` but I think the concept is the same.

**Comment by kuon** [comment]
Yeah, look similar, I wonder what @mdgriffith has to say about it and its inclusion in the lib.

**Comment by mdgriffith** [comment]
I'm hesitant to add something like `mixin` because my concern is that people will just use it to recreate stylesheets, which I'm not entirely sure is the best way to organize everything.  However, I see where the desire is coming from.

The tooltip example is nice and concrete, though more recent work on how pseudoclasses are handled should make that case really easy 😄 

I'd love to see more concrete cases where this shows up if you have them.

**Comment by kuon** [comment]
I understand your concern @mdgriffith but I think creating reusable style block with functions is something we should support. We should not forget how designers work and that many of them come from CSS. Being able to create mixins would not encourage the "stylesheet behavior" but would provide support for easier migration or complex edge cases, like the tooltip example above.

I think calling the function `merge` would avoid creating a whole `mixin` concept while solving the issue.

**Comment by mdgriffith** [comment]
So, the problem with a `mixin/merge` is that now your designer/developer has to decide which properties should be bound together.  Even though that seems straightforward, I've found that it's easy to get wrong.  

An example is having a "class" that specifies both padding and text color.  Now, these two properties are bound to each other, maybe under a name like `buttonStyle`.    But is it true that all your buttons really want that padding and that text color?  What about a large button?  Or a dark button?  Do we make separate classes for each?  Doesn't sound too crazy, but I don't think it leads to a fantastic design system on a larger scale.  

I'm still processing what the better solution would be, but currently I'm contemplating a more specific type of merge, possibly under the module `Palette`.

The idea would be to group properties that "care about" each other such as groups for colors, fonts, and spacings.

```elm
module Element.Palette exposing (colors, space, fonts)

colors : { background : Color, text : Color, border : Color } -> Element.Attribute msg

space : { padding, spacing } -> Element.Attribute msg

font : { size, family, lineHeight, } -> Element.Attribute msg

```

Essentially this would allow you to create a design vocabulary where you could do something like 

```elm
el [ Palette.colors dark, Palette.space small, Palette.font title ] (text ....)
```

I think this organization could be really nice for getting your brain to think about design systems instead of just the technical challenge you're looking at.  It also leads to a pretty predictable organization, which is nice.

**Comment by kuon** [comment]
I get the Palette idea, but I think it is really bound to the context of what attributes should be grouped. For example padding and font size are linked if you are writing typography "mixins", and if you will end with:

```
[
if big then
 Palette.space bigSpace
else
 Palette.space smallSpace
,
if big then
  Palette.font bigFont
else
  Palette.font smallFont
]
```

While I'd like those properties to be linked together. You say that the developer has to decide, but I think it is a decision for the developer, I understand it is not an easy one, and that it can go wrong rapidly, but if you don't provide the ability, we will use helpers to mimic the behavior and end with harder to read code. 

I am not saying it should be the base of the design system, but it is an API that is desirable. It can come with a warning in the documentation.

**Comment by mdgriffith** [comment]
My initial thought for that example is why parameterize a view function around `big`, why not just write it out concretely?

So when you need a big button, you write:
```elm
Input.button 
     [ Palette.space bigSpace, Palette.font bigFont ]
     { .. }
```

Though I do think it's interesting to think about where `lineHeight` and `Font.size` should be.  Maybe they should be a part of the `Palette.space` :).

**Comment by kuon** [comment]
Because your UI might have user settings for button size, like mine does.

Of course I can write an helper for it:

```
button: Size -> {onPress: Maybe msg, label: String} -> Element msg
button size {onPress, label} =
  case size of
    Big ->...
    Small -> ...
```

But then it becomes harder to manage attributes, as I might have a button with an icon or a loading spinner centered...

**Comment by mdgriffith** [comment]
For that usecase, I wonder if you'd want to just parameterize the palette itself.

Something like:

```elm

Input.button 
     [ Palette.space (space size)
     , Palette.font (fonts size)
     ]
     { .. }

```

**Comment by kuon** [comment]
For enabled/disable inputs I use the following (I use CSS naming as there are things I haven't done in SE), this covers both the button itself and the label as SE separates them:

- `background` (a gradient on enabled state should be overriden by a plain color when disabled)
- `border` (color, style and radius, I don't change width as it changes layout)
- `box-shadow`
- `outline` (easy way to get secondary border)
- `text-shadow`
- `color`
- `font-style`, `font-weight` and `font-size`, sometime even `text-decoration` or `letter-spacing`
- `opacity` (whole element opacity)

I have to say that this is the same set of properties that I also change for `:hover` state, with the addition of `transform` and `animation` (usually with a `transform`) for `:hover`.

**Comment by kuon** [comment]
I just thought of something I did in an app a while back, it's changing placeholder style when the input is disabled or focused.
